xray/core/rolling.py

```diff
@@ -0,0 +1,238 @@
+            raise ValueError('window must be > 0')
+
+        self.obj = obj
+
+        # attributes
+        self.window = window
+        self.min_periods = min_periods
+        self.center = center
+        self.dim = dim
+
+        self._axis_num = self.obj.get_axis_num(self.dim)
+
+        self._windows = None
+        self._valid_windows = None
+        self.window_indices = None
+        self.window_labels = None
+
+        self._setup_windows(min_periods=min_periods, center=center)
+
+    def __unicode__(self):
+        """ provide a nice str repr of our rolling object """
+
+        attrs = ["{k}->{v}".format(k=k, v=getattr(self, k))
+                 for k in self._attributes if getattr(self, k, None) is not None]
+        return "{klass} [{attrs}]".format(klass=self.__class__.__name__,
+                                          attrs=','.join(attrs))
+
+    @property
+    def windows(self):
+        if self._windows is None:
+            self._windows = OrderedDict(zip(self.window_labels,
+                                            self.window_indices))
+        return self._windows
+
+    def __len__(self):
+        return len(self.obj[self.dim])
+
+    def __iter__(self):
+        for (label, indices, valid) in zip(self.window_labels,
+                                           self.window_indices,
+                                           self._valid_windows):
+            if valid:
+                yield (label, self.obj.isel(**{self.dim: indices}))
+            else:
+                # return nans for this window
+                temp = self.obj.isel(**{self.dim: indices}).copy()
+                temp.values[:] = np.nan
+                yield (label, temp)
+
+    def _setup_windows(self, min_periods=None, center=False):
+        """
+        Find the indicies and labels for each window
+        """
+        from .dataarray import DataArray
+
+        window = int(self.window)
+        if min_periods is None:
+            min_periods = window
+
+        dim_size = self.obj[self.dim].size
+
+        stops = np.arange(dim_size) + 1
+        starts = np.maximum(stops - window, 0)
+
+        if min_periods > 1:
+            self._valid_windows = stops - starts >= min_periods
+        else:
+            # No invalid windows
+            self._valid_windows = np.ones(dim_size, dtype=bool)
+
+        self.window_indices = [slice(start, stop)
+                               for start, stop in zip(starts, stops)]
+
+        # Get window labels
+        if not center:
+            # get labels on the left (these are the same as the current coords)
+            self.window_labels = self.obj[self.dim]
+        else:
+            coord = self.obj[self.dim].copy()
+            orig_dtype = coord.dtype
+
+            # mean does not work on datetime64 objects
+            if np.issubdtype(coord.dtype, np.datetime64):
+                coord = coord.astype('int64')
+
+            labels = [coord[s].mean().item() for s in self.window_indices]
```

Comment:
I think this also gets the convention consistent with pandas, though it's certainly worth writing those consistency checks to verify:
```
In [29]: pd.rolling_mean(s, 4, center=True)
Out[29]:
0    NaN
1    NaN
2    1.5
3    2.5
4    3.5
5    NaN
dtype: float64
```